dunai-frp-bearriver/src/FRP/BearRiver.hs

```diff
@@ -617,3 +617,109 @@ replaceOnce a = dSwitch (arr $ const (a, Event ())) (const $ arr id)
 
 -- ** Tuples
 dup  x     = (x,x)
+
+-- * Monad transformations
+
+-- | Lift inner monadic actions in SFs
+liftTransSF :: (MonadTrans t, Monad m, Monad (t m)) => SF m a b -> SF (t m) a b
+liftTransSF sf = readerS (liftTransS (runReaderS sf))
+
+-- ** Reader layer
+
+-- | Build an 'SF' in the 'Reader' monad from one that takes the reader
+-- environment as an extra input. This is the opposite of 'runReaderSF'.
+readerSF :: Monad m => SF m (r, a) b -> SF (ReaderT r m) a b
+readerSF sf = morphS commuteReaderReader (readerS sf)
+
+-- | Build an 'SF' that takes an environment as an extra input from one on the
+-- 'Reader' monad. This is the opposite of 'readerSF'.
+runReaderSF :: Monad m => SF (ReaderT r m) a b -> SF m (r, a) b
+runReaderSF sf = runReaderS (morphS commuteReaderReader sf)
+
+-- | Build an 'SF' /function/ that takes a fixed environment as additional
+-- input, from an SF in the 'Reader' monad
+runReaderSF_ :: Monad m => r -> SF (ReaderT r m) a b -> SF m a b
+runReaderSF_ r sf = runReaderS_ (morphS commuteReaderReader sf) r
+
+-- | Lift an 'SF' in a certain monad to one in the same monad transformed
+-- with 'ReaderT'
+liftReaderSF :: Monad m => SF m a b -> SF (ReaderT r m) a b
+liftReaderSF sf = morphS commuteReaderReader (liftTransS sf)
```

Comment:
See https://github.com/turion/rhine/blob/686632679915c0415a619ff7ff5ab557fc9c08f4/rhine/src/FRP/Rhine/ClSF/Core.hs#L80 and https://github.com/turion/rhine/blob/686632679915c0415a619ff7ff5ab557fc9c08f4/rhine/src/FRP/Rhine/ClSF/Core.hs#L68. You shouldn't have to implement a lifting function for every transformer.

Reply:
I just based this on `FRP.Rhine.ClSF.Random` and wasn't aware of a different technique. Can you describe what you're hoping to see?

Reply:
I expect there is a function `liftSF :: (Monad m, MonadTrans t) => SF m a b -> SF (t m) a b` which should be implemented just like https://github.com/turion/rhine/blob/686632679915c0415a619ff7ff5ab557fc9c08f4/rhine/src/FRP/Rhine/ClSF/Core.hs#L68. This function can replace `liftReaderSF`, `liftStateSF`, and the other specialised lifting functions.

Reply:
Oh you're right. I can use `liftTransSF` anywhere I use `liftReaderSF`, `liftStateSF` or `liftWriterSF`. I removed those and re-implemented `liftTransSF` as `liftSF` using a new function `hoistSF`.